pkcs11/src/new/mod.rs

```diff
@@ -359,8 +362,42 @@ mod tests {
 
         // delete key
         session.destroy_object(is_it_the_public_key).unwrap();
+    }
+
+    #[test]
+    fn login_feast() {
+        let pkcs11 = Pkcs11::new("/usr/local/lib/softhsm/libsofthsm2.so").unwrap();
+        const SESSIONS: usize = 100;
 
-        // log out
-        session.logout().unwrap();
+        // initialize the library
+        pkcs11.initialize(CInitializeArgs::OsThreads).unwrap();
+
+        // find a slot, get the first one
+        let slot = pkcs11.get_slots_with_token().unwrap().remove(0);
+
+        // set flags
+        let mut flags = Flags::new();
+        flags.set_rw_session(true).set_serial_session(true);
+
+        let pkcs11 = Arc::from(pkcs11);
+        let mut threads = Vec::new();
+
+        for _ in 0..SESSIONS {
+            let pkcs11 = pkcs11.clone();
+            threads.push(thread::spawn(move || {
+                let pin = String::from("123456");
+                let mut session = pkcs11.open_session_no_callback(slot, flags).unwrap();
+                session.login(UserType::User, &pin).unwrap();
+                session.login(UserType::User, &pin).unwrap();
+                session.login(UserType::User, &pin).unwrap();
```

Comment:
I suggest adding a `session.login` with a wrong PIN after other sessions are already logged in - that way you can check that a wrong PIN is not allowed.

Reply:
Good idea!

Reply:
Hmm so this is interesting, I actually modified the test to be:
```rust
    #[test]
    fn login_feast() {
        let pkcs11 = Pkcs11::new("/usr/local/lib/softhsm/libsofthsm2.so").unwrap();
        const SESSIONS: usize = 1;

        // initialize the library
        pkcs11.initialize(CInitializeArgs::OsThreads).unwrap();

        // find a slot, get the first one
        let slot = pkcs11.get_slots_with_token().unwrap().remove(0);

        // set flags
        let mut flags = Flags::new();
        flags.set_rw_session(true).set_serial_session(true);

        let pkcs11 = Arc::from(pkcs11);
        let mut threads = Vec::new();

        for _ in 0..SESSIONS {
            let pkcs11 = pkcs11.clone();
            threads.push(thread::spawn(move || {
                let pin = String::from("123456");
                let bad_pin = String::from("TOTO IS GOOD");
                let session = pkcs11.open_session_no_callback(slot, flags).unwrap();
                session.login(UserType::User, &pin).unwrap();
                session.login(UserType::User, &pin).unwrap();
                let err = session.login(UserType::User, &bad_pin).unwrap_err();
                if let Error::Pkcs11(rv_err) = dbg!(err) {
                    assert_eq!(rv_err, RvError::PinIncorrect);
                } else {
                    panic!("Expected an incorrect pin.");
                }
                session.logout().unwrap();
                session.logout().unwrap();
                session.logout().unwrap();
            }));
        }

        for thread in threads {
            thread.join().unwrap();
        }
    }
```
but the `unwrap_err` panics `thread '<unnamed>' panicked at 'called `Result::unwrap_err()` on an `Ok` value: ()'`.

The reason is that even though we call `C_Login` to check if the pin is correct, the `CKR_USER_ALREADY_LOGGED_IN` error seems to be prioriaty on the `CKR_PIN_INCORRECT` one. I believe that according to the spec of the `C_Login` method, this is normal behaviour and the `CKR_PIN_INCORECT` seems to be one of the last error returned:

> When C_Login returns, whatever authentication method supported by the token will have been performed; a return value of CKR_OK means that the user was successfully authenticated, and a return value of CKR_PIN_INCORRECT means that the user was denied access.

That means that with the current logic we can not check in the pin is correct when another session is already logged in.
The logic is then also a bit more insecure because a malicious session could wait for the user to be logged in and then call the `login` function to then have unlimited login access, even with a wrong pin. Although I am not sure if one application would let ownership of some sessions to malicious entities.

We can either ignore this threat or change the implementation of login to not do anything special and leave it to the user to implement their own logic.